pkgs/jni/lib/src/jreference.dart

```diff
@@ -62,45 +67,76 @@ abstract final class JReference {
   ///
   /// Further uses of this object will throw [UseAfterReleaseError].
   void release() {
-    setAsReleased();
+    _setAsReleased();
     _deleteReference();
   }
 
   void _deleteReference();
+
+  void _setAsReleased();
 }
 
 /// A managed JNI global reference.
 ///
 /// Uses a [NativeFinalizer] to delete the JNI global reference when finalized.
-final class JGlobalReference extends JReference implements Finalizable {
-  static final _finalizer =
-      NativeFinalizer(Jni.env.ptr.ref.DeleteGlobalRef.cast());
-
-  JGlobalReference(super._reference) {
-    _finalizer.attach(this, _pointer, detach: this);
+@pragma('vm:deeply-immutable')
+final class JGlobalReference extends JReference {
+  /// The finalizable handle that deletes [_JFinalizable.pointer].
+  final Dart_FinalizableHandle _jobjectFinalizableHandle;
+  final Pointer<Bool> _isReleased;
+
+  JGlobalReference._(
+      super._finalizable, this._jobjectFinalizableHandle, this._isReleased);
+
+  factory JGlobalReference(Pointer<Void> pointer) {
+    final finalizable = _JFinalizable(pointer);
+    final isReleased = calloc<Bool>();
+    final jobjectFinalizableHandle =
+        ProtectedJniExtensions.newJObjectFinalizableHandle(
+            finalizable, finalizable.pointer, JObjectRefType.JNIGlobalRefType);
+    ProtectedJniExtensions.newBooleanFinalizableHandle(finalizable, isReleased);
+    return JGlobalReference._(
+        finalizable, jobjectFinalizableHandle, isReleased);
   }
 
   @override
   bool get isNull => pointer == nullptr;
 
+  @override
+  void _setAsReleased() {
+    if (isReleased) {
+      throw DoubleReleaseError();
+    }
+    _isReleased.value = true;
```

Comment:
Shouldn't this be a compare and swap? This object is shared across multiple isolates. if you call delete twice concurrently this can double release.

bool oldValue = compareAndSwap(pointer, bool newValue);
if(oldValue) throw DoubleReleaseError();
deleteFinalizableHandle(...);